android/engine/src/main/java/org/smartregister/fhircore/engine/rulesengine/RulesExecutor.kt

```diff
@@ -0,0 +1,208 @@
+              resource = resource,
+              relatedResourceKey = id ?: resourceType.name,
+              referenceFhirPathExpression = fhirPathExpression,
+              relatedResourcesMap = relatedResourcesMap
+            )
+        }
+
+      val listComputedValuesMap =
+        computeRules(
+          ruleConfigs = ruleConfigs,
+          baseResource = resource,
+          relatedResourcesMap = listItemRelatedResources
+        )
+
+      // LIST view should reuse the previously computed values
+      ResourceData(
+        baseResourceId = resource.logicalId.extractLogicalIdUuid(),
+        baseResourceType = resource.resourceType,
+        computedValuesMap = computedValuesMap.plus(listComputedValuesMap)
+      )
+    }
+
+  /**
+   * This function returns a list of filtered resources. The required list is obtained from
+   * [relatedResourceMap], then a filter is applied based on the condition returned from the
+   * extraction of the [ListResource] conditional FHIR path expression
+   */
+  private fun filteredListResources(
+    relatedResourceMap: MutableMap<String, MutableList<Resource>>,
+    listResource: ListResource
+  ): MutableList<Resource> {
+    val relatedResourceKey = listResource.relatedResourceId ?: listResource.resourceType.name
+    val newListRelatedResources = relatedResourceMap[relatedResourceKey]
+
+    // conditionalFhirPath expression e.g. "Task.status == 'ready'" to filter tasks that are due
+    if (newListRelatedResources != null &&
+        !listResource.conditionalFhirPathExpression.isNullOrEmpty()
+    ) {
+      return rulesFactory
+        .rulesEngineService
+        .filterResources(
+          resources = newListRelatedResources,
+          fhirPathExpression = listResource.conditionalFhirPathExpression
+        )
+        .toMutableList()
+    }
+
+    return newListRelatedResources ?: mutableListOf()
+  }
+}
+
+/**
+ * This function creates a map of resource config Id ( or resource type if the id is not configured)
+ * against [Resource] from a list of nested [RepositoryResourceData].
+ *
+ * Example: A list of [RepositoryResourceData] with Patient as its base resource and two nested
+ * [RepositoryResourceData] of resource type Condition & CarePlan returns:
+ * ```
+ * {
+ * "Patient" -> [Patient],
+ * "Condition" -> [Condition],
+ * "CarePlan" -> [CarePlan]
+ * }
+ * ```
+ *
+ * NOTE: [RepositoryResourceData] are represented as tree however they grouped by their resource
+ * config Id ( or resource type if the id is not configured) as key and value as list of [Resource]
+ * s in the map.
+ */
+fun LinkedList<RepositoryResourceData>.createRelatedResourcesMap():
+  MutableMap<String, MutableList<Resource>> {
+  val relatedResourcesMap = mutableMapOf<String, MutableList<Resource>>()
+  while (this.isNotEmpty()) {
+    val relatedResourceData = this.removeFirst()
+    relatedResourcesMap
+      .getOrPut(relatedResourceData.configId ?: relatedResourceData.resource.resourceType.name) {
```

Comment:
kinda the same issue as on line 103

Reply:
this is interesting, not sure how expensive each iteration in this loop is, if it's expensive enough worth considering using [parallelStream](https://blogs.oracle.com/javamagazine/post/java-parallel-streams-performance-benchmark) here. You would have to restructure to have it be tiered, like parallel process one level, to build next level, then parallel process next level etc, but that should be doable

Reply:
Same as above on _Duplicate Exceptions_

Reply:
yea let's throw, or maybe this calls for a helper that we call in both places, the code looks similar, if it's the same a helper makes sense that also handles the throwing